Refactor GroovycWrapper into two parts

natgabb opened the pull request:
- One part is the TreeParser and takes care of parsing symbols.
- The other part is the WorkspaceCompiler which takes care of compiling the code.
- The CompilerWrapperProvider is gone and now the CompilerWrapper is just part of the shared state.
- New packages names: ls.groovy.api, ls.groovy.compilation, ls.groovy.services

Closes #84

Commit:
Refactor GroovycWrapper into two parts

- One part is the TreeParser and takes care of parsing symbols.
- The other part is the WorkspaceCompiler which takes care of compiling the code.
- The CompilerWrapperProvider is gone and now the CompilerWrapper is just part of the shared config.
- New packages names: ls.api, ls.groovy.compilation, ls.groovy.services

Closes #84

natgabb commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyTreeParser.java:
```diff
@@ -66,163 +53,107 @@
      */
-    public static GroovycWrapper of(Path targetDirectory, Path workspaceRoot, Path changedFilesRoot) {
-        checkNotNull(targetDirectory, "targetDirectory must not be null");
+    public static GroovyTreeParser of(CompilationUnitProvider unitProvider, Path workspaceRoot, Path changedFilesRoot) {
```
`GroovyTreeParser` shouldn't need `changedFilesRoot`. I need to figure out another way to resolve uri in `getWorkspaceUri`

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java:
```diff
@@ -31,13 +32,25 @@
 
 public final class GroovyLanguageServerConfig implements LanguageServerConfig {
```
Can we rename this class to `GroovyLanguageServerSharedObjects` or something like that
Same with the interface

natgabb replied:
How about `GroovyLanguageServerState` ?

jiahuijiang commented:
looks good in general :)

Commit:
Add the WorkspaceUriSupplier

Commit:
Move tests to be in the right packages and rename UriSupplier

Commit:
Rename LanguageServerConfig to LanguageServerState

natgabb commented:
@gracew can you please take a look when you get a chance :)

natgabb commented:
I still need to refactor the tests into two separate files but I think it might be easier to do in a follow up PR.

Commit:
Change packages

- ls.api -> ls.groovy.api
- ls.server -> ls.groovy.server
- ls.util -> ls.groovy.util

Commit:
Add WorkspaceUriSupplierTest

Commit:
Have GroovyWorkspaceCompiler implement Supplier<CompilationUnit>
